### PR TITLE
Add COPC reading support

### PR DESCRIPTION
**Pull request by tmontaigu**
This adds the `CopcReader` class to laspy
that understands the COPC LAZ format, enabling to
do spatial and / or LOD based queries.

The class require `lazrs` to work,
and optionaly, if `requests` is installed
COPC over HTTP is supported.

### CODE REVIEW

File: laspy/copc.py [hunk 1]
```diff
@@ -0,0 +1,863 @@
...
+    def run(self) -> None:
+        with HttpRangeStream(self.url) as http_reader:
+            while not self.query_queue.empty():
+                offset, size = self.query_queue.get()
+                try:
+                    http_reader.seek(offset)
+                    data = http_reader.read(size)
+                    self.result_queue.put((data, offset))
+                except Exception as e:
+                    print("AYAYA")
```
Comment (tmontaigu) [inline laspy/copc.py:488]: 🥵

File: laspy/copc.py [hunk 2]
```diff
@@ -0,0 +1,863 @@
...
+                num_compressed_group_bytes += node.byte_size
+                num_points += node.point_count
+
+            num_compressed_bytes += num_compressed_group_bytes
+            byte_queries.append((group[0].offset, num_compressed_group_bytes))
+
+        compressed_bytes = bytearray(num_compressed_bytes)
+
+        if isinstance(self.source, HttpRangeStream):
+            http_queue_strategy(
```
Comment (tmontaigu) [inline laspy/copc.py:838]: add semi-hidden parameter to chose the strategy

File: laspy/copc.py [hunk 3]
```diff
@@ -0,0 +1,863 @@
...
+            for offset, size in byte_queries:
+                self.source.seek(offset)
+                cc = citer.next(size)
+                self.source.readinto(cc)
+        else:
+            citer = ChunkIter(np.frombuffer(compressed_bytes, np.uint8))
+            for offset, size in byte_queries:
+                self.source.seek(offset)
+                cc = citer.next(size)
+                cc[:] = np.frombuffer(self.source.read(size), np.uint8)
```
Comment (tmontaigu) [inline laspy/copc.py:854]: could be avoid all those `np.from_buffer` cast by using [memoryviews](https://docs.python.org/3/library/stdtypes.html#memoryview)

File: laspy/copc.py [hunk 4]
```diff
@@ -0,0 +1,863 @@
...
+        self.http_num_threads = http_num_threads
+
+        # The idea here is to limit as much as possible the number of read
+        # to a slow external ressource like an HTTP endpoint
+        # by pre-fetching in memory what we need.
+        # COPC files _must_ be 1.4 LAS
+        header_bytes = stream.read(LAS_HEADERS_SIZE["1.4"])
+        offset_to_point_data = struct.unpack_from("<I", header_bytes, offset=96)[0]
+        rest = stream.read(offset_to_point_data - len(header_bytes))
+        header_source = io.BytesIO(header_bytes + rest)
```
Comment (tmontaigu) [inline laspy/copc.py:628]: That should probably what the implementation of `LasHeader.read_from` should do by default

File: laspy/copc.py [hunk 5]
```diff
@@ -0,0 +1,863 @@
...
+    root_node.key.level = 0
+
+    nodes_to_load: List[OctreeNode] = [root_node]
+    while nodes_to_load:
+        current_node = nodes_to_load.pop()
+
+        try:
+            entry = root_page.entries[current_node.key]
+        except KeyError:
+            continue
```
Comment (tmontaigu) [inline laspy/copc.py:380]: We should probably also delete from parent here

**Comment by hobu** [comment]
In addition to filtering by a bounds, it would also be really useful to filter by a resolution or tree level.  PDAL has some examples of this resolution-based filtering at https://github.com/PDAL/PDAL/tree/master/io/private/copc and https://github.com/PDAL/PDAL/blob/master/io/CopcReader.cpp#L241

**Comment by tmontaigu** [comment]
> In addition to filtering by a bounds, it would also be really useful to filter by a resolution or tree level. PDAL has some examples of this resolution-based filtering at https://github.com/PDAL/PDAL/tree/master/io/private/copc and https://github.com/PDAL/PDAL/blob/master/io/CopcReader.cpp#L241

The query function is defined as

```Python
    def query(
        self,
        bounds: Optional[Bounds] = None,
        level: Optional[Union[int, range]] = None,
    ) -> ScaleAwarePointRecord:
```

So you can filter by bounds or level or both at the same time , some wrapper functions also exists
```Python
    def spatial_query(self, bounds: Bounds) -> ScaleAwarePointRecord:
        return self.query(bounds=bounds, level=None)

    def level_query(self, level: Union[int, range]) -> ScaleAwarePointRecord:
        return self.query(bounds=None, level=level)
```

**Comment by hobu** [comment]
> Ok I got confused, resolution != level

Yes. They're directly related, but we have found with EPT and COPC it is much easier to say "select everything at or below 3m" instead of having to calculate based on the dimensions of the octree what level would correspond to everything 3m or below.

**Comment by hobu** [comment]
Some more files to play with



https://viewer.copc.io/?copc=https://s3.amazonaws.com/hobu-lidar/montreal-2015.copc.laz
https://viewer.copc.io/?copc=https://s3.amazonaws.com/hobu-lidar/melbourne-2018.copc.laz

**Comment by tmontaigu** [comment]
resolution was added as a query parameter

**Comment by hobu** [comment]
I attempted to test this but got the following:
> AttributeError: module 'lazrs' has no attribute 'decompress_points_with_chunk_table'

I presume this is due to lazrs not being released with chunk table support quite yet?

**Comment by tmontaigu** [comment]
Yes, I plan on releasing lazrs later this week, after that this PR will be testable easily

**Comment by tmontaigu** [comment]
The new version of lazrs was pushed to pypi so it should be easier to test now

**Comment by hobu** [comment]
OSX arm wheels for lazrs didn't post https://pypi.org/project/lazrs/#files

**Comment by hobu** [comment]
I installed my own lazrs. 

It seems a bit slower than PDAL while fetching over the network. Are you reading the entire octree down to the client?

```
time pdal info --all https://s3.amazonaws.com/hobu-lidar/melbourne-2018.copc.laz   
1.59s user 1.48s system 110% cpu 2.786 total
```

```
time python copc.py https://s3.amazonaws.com/hobu-lidar/melbourne-2018.copc.laz
copc_reader_ready for https://s3.amazonaws.com/hobu-lidar/melbourne-2018.copc.laz
Spatial Query gave: <ScaleAwarePointRecord(fmt: <PointFormat(7, 0 bytes of extra dims)>, len: 77486, point size: 36)>
0.021911293026868
Spatial Query gave: <ScaleAwarePointRecord(fmt: <PointFormat(7, 0 bytes of extra dims)>, len: 101208, point size: 36)>
0.028619339553767862
It took: 10.786364793777466  seconds
python copc.py https://s3.amazonaws.com/hobu-lidar/melbourne-2018.copc.laz  2.86s user 2.64s system 50% cpu 10.950 total
```

**Comment by tmontaigu** [comment]
> OSX arm wheels for lazrs didn't post https://pypi.org/project/lazrs/#files 

Maybe you have a arm mac I have not posted these wheels yet ?

> I installed my own lazrs. 
> 
> 
> 
> It seems a bit slower than PDAL while fetching over the network. Are you reading the entire octree down to the client?

Yes the entire octree is fetched at the begnning. 
The idea was that when doing multiple queries on the same ressource, we would not have to refetch the octree info.
I'll change it to only fetch progressively the node informations and see how much it improves.

The `pdal info` only reads the header and does not fetch any octree info nor points ?

The python example fetches the whole octree info and then queries 2 quarters of the bounds and decompress them

**Comment by tmontaigu** [comment]
I just pushed changes that makes it so that the octree hierarchies are fetched when actually needed (and not fetch every thing on init like before)

**Comment by hobu** [comment]
> I just pushed changes that makes it so that the octree hierarchies are fetched when actually needed (and not fetch every thing on init like before)

Confirmed and behaving as I would expect now. 

>  I have not posted these wheels yet

I didn't realize these weren't automated. https://github.com/laz-rs/laz-rs-python/issues/9 implied they were, but maybe that isn't quite done yet.

**Comment by hobu** [comment]
Sorry, my `pdal info` call was this:

```
time pdal info --all https://s3.amazonaws.com/hobu-lidar/melbourne-2018.copc.laz --readers.copc.resolution=20.0
```

That reads across the entire domain and fetches points to a 20m resolution (376303 points)

**Comment by tmontaigu** [comment]
> Confirmed and behaving as I would expect now.

Ok nice


> I didn't realize these weren't automated. [laz-rs/laz-rs-python#9](https://github.com/laz-rs/laz-rs-python/issues/9) implied they were, but maybe that isn't quite done yet.

Yeah its a thing that I want to do, it should not be hard, but CI is not the most interesting thing to do in projects 😅

**Pull request merged**